### PR TITLE
Fixes #94 (`selected` does not update when `attr-for-selected` changes.)

### LINKED ISSUE #94: `selected` does not update when `attr-for-selected` changes.
When `attr-for-selected` changes, `selected` should update it's value to the value of the new attribute referenced by `attr-for-selected` of the currently selected item. [Here's a demo.](http://jsbin.com/xedazaweku/edit?html,console,output) Select an item and then click the 'attr-for-selected = data-y' button - `selected` doesn't change.

### PR DESCRIPTION
**Pull request by MeTaNoV**
fixes Issue #94  when attrForSelected change without reflecting the change to selected both in selectable and multi-selectable

### CODE REVIEW

File: iron-selectable.html [hunk 1]
```diff
@@ -148,7 +149,7 @@
       this._observer = this._observeItems(this);
       this._updateItems();
       if (!this._shouldUpdateSelection) {
-        this._updateSelected(this.attrForSelected,this.selected)
+        this._updateSelected(this.selected);
```
Comment (bicknellr) [inline iron-selectable.html]: We might as well remove the `this.selected` argument as well given that `_updateSelected` does nothing with it.

File: iron-selectable.html [hunk 2]
```diff
@@ -241,6 +242,12 @@
       this._setItems(nodes);
     },
 
+    _updateAttrForSelected: function() {
+      if (this._shouldUpdateSelection) {
+        this.selected = this._indexToValue(this.indexOf(this.selectedItem));
```
Comment (bicknellr) [inline iron-selectable.html]: Nice, I was looking at this issue earlier and was trying to do this by caching the old value of `attrForSelected`. This is much cleaner.

File: iron-multi-selectable.html [hunk 3]
```diff
@@ -77,6 +78,18 @@
...
     },
 
+    _updateAttrForSelected: function() {
+      if (this._shouldUpdateSelection) {
+        if (this.multi) {
+          this.selectedValues = this.selectedItems.map(function(selectedItem) {
+            return this._indexToValue(this.indexOf(selectedItem));        
+          }, this);
+        } else {
+          this.selected = this._indexToValue(this.indexOf(this.selectedItem));
```
Comment (bicknellr) [inline iron-multi-selectable.html]: I think it would be better to call `Polymer.IronSelectableBehavior._updateAttrForSelected)` here instead of duplicating this line.

File: test/selected-attribute.html [hunk 4]
```diff
@@ -66,6 +76,24 @@
...
+
+      var s;
+
+      setup(function () {
+        s = fixture('test-attr-change');
+      });
+
+      test('changing selectedAttribute', function() {
+        s.attrForSelected = 'data-y';
+        s.async(function() {
```
Comment (bicknellr) [inline test/selected-attribute.html]: Does this really need to be in a call to `async`?

File: iron-multi-selectable.html (hunk 3 above)
Comment (bicknellr) [inline iron-multi-selectable.html]: Maybe flip this so that it's
```
if !this.multi
  call super
else if this._shouldUpdateSelection
  do selectedValues mapping
```
.. then, the _shouldUpdateSelection test only happens once when multiple selection is disabled.

**Comment by bicknellr** [comment]
Thanks for submitting this! I left a few comments for you to look through. Also, could you fill out the tests a bit more? For example, there aren't any tests for when multiple selection is enabled. Otherwise, it looks good.

**Comment by bicknellr** [comment]
Also, way too many of my sentences are structured like this.

File: iron-selectable.html (hunk 1 above)
Comment (MeTaNoV) [inline iron-selectable.html]: jup, that's right! I feel always uncomfortable with the observation definition, where you use your argument to trigger the function execution knowing you can always access the property inside your element... :)

File: iron-selectable.html (hunk 2 above)
Comment (MeTaNoV) [inline iron-selectable.html]: I though about this solution to keep only one observer, but, yeah, separation of concerns is indeed cleaner

File: iron-multi-selectable.html (hunk 3 above)
Comment (MeTaNoV) [inline iron-multi-selectable.html]: I tried to call Polymer.IronSelectableBehavior._updateAttrForSelected(), but 'this' is a different one which does not contain the selected property set correctly... I'll try again...

**Comment by MeTaNoV** [comment]
I'll add a test for multi, oki doki. What about your last remark? :) Like common answer to a PR? then use a template! :)

File: test/selected-attribute.html (hunk 4 above)
Comment (MeTaNoV) [inline test/selected-attribute.html]: I tried w/o in the first place, but it seems the value is not updated properly "in time" in this case... Do you have a better solution? I had also such syncing issue while developing an iron-swipeable-pages element where selection or change of class were not propagating in time for the code to execute properly w/o some call to async()...

**Comment by MeTaNoV** [comment]
Hi Russel ( @bicknellr ), thanks for the follow-up, I took your feedback into account, apart the `async()` usage which make the test fail if not used. Would be interested to know if such syncing issue could be avoided...

File: iron-multi-selectable.html [hunk 5]
```diff
@@ -77,6 +78,16 @@
         (this.selectedValues != null && this.selectedValues.length);
     },
 
+    _updateAttrForSelected: function() {
+      if (!this.multi) {
+        Polymer.IronSelectableBehavior._updateAttrForSelected();
```
Comment (bicknellr) [inline iron-multi-selectable.html]: `Polymer.IronSelectableBehavior._updateAttrForSelected.apply(this);`

File: test/selected-attribute.html (hunk 4 above)
Comment (bicknellr) [inline test/selected-attribute.html]: Just left a comment about using `apply` in IronMultiSelectable behavior. I think that might be the reason. (Or was this here before that?)

File: iron-multi-selectable.html (hunk 5 above)
Comment (MeTaNoV) [inline iron-multi-selectable.html]: sadly it doesn't help... same behavior, only updated using async()...

File: test/selected-attribute.html (hunk 4 above)
Comment (MeTaNoV) [inline test/selected-attribute.html]: sadly same behavior...

File: iron-multi-selectable.html (hunk 5 above)
Comment (bicknellr) [inline iron-multi-selectable.html]: You'll need to call it this way anyways though; otherwise `Polymer.IronSelectableBehavior._updateAttrForSelected` won't be called in the context of this object, but rather `Polymer.IronSelectableBehavior` itself.

File: test/selected-attribute.html [hunk 6]
```diff
@@ -43,6 +43,28 @@
...
+        <div data-x="x-3" data-y="y-3">3</div>
+      </iron-selector>
+    </template>
+  </test-fixture>
+
+  <test-fixture id="test-attr-change-multi">
+    <template>
+      <iron-selector multi id="selector" 
+                     attr-for-selected="data-x" 
+                     selected-values='["x-1",""x-2"]'>
```
Comment (bicknellr) [inline test/selected-attribute.html]: one too many `"` here

Comment (bicknellr) [inline test/selected-attribute.html]: I'd also say flip them for consistency.

Comment (bicknellr) [inline test/selected-attribute.html]: Oh wait, the JSON parser...

File: iron-multi-selectable.html (hunk 5 above)
Comment (MeTaNoV) [inline iron-multi-selectable.html]: jup jup

File: test/selected-attribute.html (hunk 6 above)
Comment (MeTaNoV) [inline test/selected-attribute.html]: I'll fix the doubled, but yeah, JSON format

File: test/selected-attribute.html [hunk 7]
```diff
@@ -66,6 +88,44 @@
...
 
+    suite('changing attrForSelected', function() {
+
+      var s;
+
+      setup(function () {
+        s = fixture('test-attr-change');
+      });
+
+      test('changing selectedAttribute', function() {
```
Comment (bicknellr) [inline test/selected-attribute.html]: Also, `done` is passed as an argument to the test.

Comment (MeTaNoV) [inline test/selected-attribute.html]: done! :)

**Comment by MeTaNoV** [comment]
@bicknellr sorry for the delay... Took into account your last remark. The `async` usage being the last question to be solved...

File: test/selected-attribute.html (hunk 6 above)
Comment (bicknellr) [inline test/selected-attribute.html]: (this still needs to be fixed btw)

**Comment by bicknellr** [comment]
I think I found out what's going on here after running into the same issue lately. There's a problem with the current implementation of `Polymer.Base.async` playing poorly with anything using a MutationObserver. Particularly, `async` uses its own MutationObserver to get microtask timing and flushes it's queue of async callbacks all at once and since `async`'s MutationObserver is pretty much always registered before anything else (when Polymer itself loads) it always runs before any other MutationObservers. So, first register + batching = all `async` callbacks always run before any MutationObserver callbacks triggered in the same stack. There's a function `Polymer.dom.flush();` which force any updates queued by shady DOM to be reflected immediately. I believe it ties into `observeNodes` (Polymer's shady DOM MutationObserver polyfill-ish thing) and will cause them to flush also.

tl;dr Put `Polymer.dom.flush();` between the time the fixtures are created and when the assertion is made and these seem to work. I'll ask around to make sure this is the right idea but `wct`-ing with all the usual browsers seems to work so I'm pretty confident.

**Comment by bicknellr** [comment]
Also, when `attrForSelected` changes and (one of) the selected node(s) doesn't have that key, what happens? I'm worried that it'll add `undefined` to the set of selected values and then select all nodes without that key. Maybe nodes without the newly chosen key should be filtered out before the values are mapped?

**Comment by MeTaNoV** [comment]
good to go! :)

File: iron-multi-selectable.html [hunk 8]
```diff
@@ -77,6 +77,18 @@
...
     },
 
+    _updateAttrForSelected: function() {
+      if (!this.multi) {
+        Polymer.IronSelectableBehavior._updateAttrForSelected.apply(this);
+      } else if (this._shouldUpdateSelection) {
+        this.selectedValues = this.selectedItems.map(function(selectedItem) {
+          return this._indexToValue(this.indexOf(selectedItem));        
+        }, this).filter(function(unfilteredValue) {
+          return unfilteredValue != null;
```
Comment (bicknellr) [inline iron-multi-selectable.html]: `!==`, unfortunately `0 != null` is true

**Comment by bicknellr** [comment]
Left a comment; otherwise, I think this is ready!

File: iron-multi-selectable.html (hunk 8 above)
Comment (MeTaNoV) [inline iron-multi-selectable.html]: there is some other '!=' to be fixed elsewhere then! :) I'll do 'em all!

**Comment by MeTaNoV** [comment]
that should be it!

File: iron-multi-selectable.html (hunk 8 above)
Comment (bicknellr) [inline iron-multi-selectable.html]: *facepalm* Sorry, I was totally wrong! Type coercion sucks.

**Comment by bicknellr** [comment]
My bad, I was under the impression that `0 != null` for some reason.. Could you drop the last commit?

**Comment by MeTaNoV** [comment]
:) done!

**Comment by bicknellr** [comment]
Awesome, thanks for all the work!
